File: src/main/java/net/gtaun/shoebill/CharsetUtils.java
package net.gtaun.shoebill;

import java.nio.ByteBuffer;
import java.nio.CharBuffer;
import java.nio.charset.Charset;
import java.nio.charset.CharsetEncoder;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class CharsetUtils
{
	private static final Map<Integer, Charset> WINDOWS_CHARSETS = new HashMap<>();
	static
	{
		WINDOWS_CHARSETS.put(862, Charset.forName("CP862"));
		WINDOWS_CHARSETS.put(874, Charset.forName("CP874"));
		WINDOWS_CHARSETS.put(932, Charset.forName("SHIFT_JIS"));
		WINDOWS_CHARSETS.put(936, Charset.forName("GBK"));
		WINDOWS_CHARSETS.put(949, Charset.forName("BIG5"));
		for (int page = 1250; page <= 1258; page++) WINDOWS_CHARSETS.put(page, Charset.forName("CP" + page));
	}


	public static Charset getCharsetByCodepage(int codepage)
	{
		return WINDOWS_CHARSETS.get(codepage);
	}

	public static List<String> splitStringByCharsetLength(String str, int maxBytes, Charset charset)
	{
		List<String> strs = new ArrayList<>();
		CharsetEncoder encoder = charset.newEncoder();

		char[] in = new char[1];
		byte[] outBuf = new byte[2];

		ByteBuffer out = ByteBuffer.wrap(outBuf);
		StringBuilder sb = new StringBuilder(maxBytes);

		for (char c : str.toCharArray())
		{
			out.clear();
			encoder.encode(CharBuffer.wrap(in), out, false);

			if (sb.length() + out.position() > maxBytes)
			{
				strs.add(sb.toString());
				sb.setLength(0);
			}

			sb.append(c);
		}

		strs.add(sb.toString());
		return strs;
	}

	private CharsetUtils()
	{

	}
}

File: src/main/java/net/gtaun/shoebill/CharsetUtils.kt
package net.gtaun.shoebill

import java.nio.ByteBuffer
import java.nio.CharBuffer
import java.nio.charset.Charset
import java.util.*

object CharsetUtils {
    private val WINDOWS_CHARSETS = HashMap<Int, Charset>()

    init {
        WINDOWS_CHARSETS.put(862, Charset.forName("CP862"))
        WINDOWS_CHARSETS.put(874, Charset.forName("CP874"))
        WINDOWS_CHARSETS.put(932, Charset.forName("SHIFT_JIS"))
        WINDOWS_CHARSETS.put(936, Charset.forName("GBK"))
        WINDOWS_CHARSETS.put(949, Charset.forName("BIG5"))
        for (page in 1250..1258) WINDOWS_CHARSETS.put(page, Charset.forName("CP" + page))
    }

    fun getCharsetByCodepage(codepage: Int): Charset? = WINDOWS_CHARSETS[codepage]

    @JvmOverloads
    fun splitStringByCharsetLength(str: String, maxBytes: Int, charset: Charset, linePrefix: String = ""): List<String> {
        val strs = ArrayList<String>()
        val encoder = charset.newEncoder()

        val `in` = CharArray(1)
        val outBuf = ByteArray(2)

        val out = ByteBuffer.wrap(outBuf)
        val sb = StringBuilder(maxBytes)

        for (c in str.toCharArray()) {
            out.clear()
            encoder.encode(CharBuffer.wrap(`in`), out, false)

            if (sb.length + out.position() > maxBytes) {
                strs.add(sb.toString())
                sb.setLength(0)
                sb.append(linePrefix)
            }

            sb.append(c)
        }

        strs.add(sb.toString())
        return strs
    }
}
